zstd/zstd.go

```diff
@@ -38,161 +35,91 @@ func (c *CompressionCodec) Name() string { return "zstd" }
-	return &writer{
-		writer: w,
-		buffer: bufferPool.Get().(*buffer),
-		level:  c.level,
-	}
-}
-
-// =============================================================================
-// The DataDog/zstd package exposes io.Writer and io.Reader implementations that
-// encode and decode streams, however there are no APIs to reuse the values like
-// other compression format have (through a Reset method usually).
-//
-// I first tried using these abstractions but the amount of state that gets
-// recreated and destroyed was so large that it was slower than using the
-// zstd.Compress and zstd.Decompress functions directly. Knowing that, I changed
-// the implementation to be more of a buffer management on top of these instead.
-// =============================================================================
+var decPool sync.Pool
 
 type reader struct {
-	reader io.Reader
-	buffer *buffer
-	offset int
+	dec *zstdlib.Decoder
+	err error
 }
 
-func (r *reader) Read(b []byte) (int, error) {
-	if err := r.decompress(); err != nil {
-		return 0, err
-	}
-
-	if r.offset >= len(r.buffer.output) {
-		return 0, io.EOF
+// Close implements the io.Closer interface.
+func (r *reader) Close() error {
+	if r.err != io.ErrClosedPipe {
+		r.err = io.ErrClosedPipe
+		decPool.Put(r.dec)
 	}
-
-	n := copy(b, r.buffer.output[r.offset:])
-	r.offset += n
-	return n, nil
+	return nil
 }
 
-func (r *reader) WriteTo(w io.Writer) (int64, error) {
-	if err := r.decompress(); err != nil {
-		return 0, err
-	}
-
-	if r.offset >= len(r.buffer.output) {
-		return 0, nil
+// Read implements the io.Reader interface.
+func (r *reader) Read(p []byte) (n int, err error) {
+	println("zstd read", len(p))
+	if r.err != nil {
+		return 0, r.err
 	}
-
-	n, err := w.Write(r.buffer.output[r.offset:])
-	r.offset += n
-	return int64(n), err
+	return r.dec.Read(p)
 }
 
-func (r *reader) Close() (err error) {
-	if b := r.buffer; b != nil {
-		r.buffer = nil
-		b.reset()
-		bufferPool.Put(b)
+// WriteTo implements the io.WriterTo interface.
+func (r *reader) WriteTo(w io.Writer) (n int64, err error) {
+	if r.err != nil {
+		return 0, r.err
 	}
-	return
+	return r.dec.WriteTo(w)
 }
 
-func (r *reader) decompress() (err error) {
-	if r.reader == nil {
-		return
-	}
-
-	b := r.buffer
-
-	if _, err = b.readFrom(r.reader); err != nil {
-		return
+// NewWriter implements the kafka.CompressionCodec interface.
+func (c *CompressionCodec) NewWriter(w io.Writer) io.WriteCloser {
+	p := new(writer)
+	if cached := encPool.Get(); cached == nil {
+		p.enc, p.err = zstdlib.NewWriter(w,
+			zstdlib.WithEncoderLevel(c.level))
+	} else {
+		p.enc = cached.(*zstdlib.Encoder)
+		p.enc.Reset(w)
 	}
-
-	r.reader = nil
-	b.output, err = zstd.Decompress(b.output[:cap(b.output)], b.input)
-	return
-}
-
-type writer struct {
-	writer io.Writer
-	buffer *buffer
-	level  int
+	return p
 }
 
-func (w *writer) Write(b []byte) (int, error) {
-	return w.buffer.write(b)
-}
+var encPool sync.Pool
```

Comment:
Do we need to set the `New` function on this pool?

Reply:
Not required. The default `nil` value helps with applying the Writer.